### ISSUE TITLE
pager="less -RF" breaks on macOS for single-page, but default pager breaks pgup/pgdn

### ISSUE THREAD

**Issue by lilyball** [sharkdp/bat#681]
The default paging behavior passes `--no-init` to `less`, which breaks both page up/down and mouse support. Unfortunately, overriding the pager with `--pager="less -RF"` breaks on macOS when the content fits within a single screen because Terminal.app's deinitialization clears the screen (or rather, it uses the alternate screen buffer for less).

I like the behavior of "don't page if the content is small enough", but the `--no-init` behavior is really aggravating (plus it screws with scrollback when I page up and down through the file because it's just dumping lots of random non-contiguous lines into scrollback). I could just give in and use `--pager="less -R"` but that just makes bat less useful when reading small files.

A way to solve this might be to give bat an option where it will skip the pager entirely if the contents can fit in one screen, making the `-F` flag to `less` unnecessary. Ideally it would match less's behavior exactly for calculating this, but being close enough is probably fine (probably more important is to make sure it never underestimates this, such that `--pager="less -FR"` is safe to use with this option so people don't have edge cases where they see no output from bat because it invoked the pager but less decided it was one screen).

**Comment by eth-p** [comment]
I would consider this to be an issue with macOS shipping with an old version of `less`. While it's not impossible to implement less's `--quit-if-one-screen` behaviour in `bat`, we would like need to buffer the output in order to determine whether or not to use the pager. It also comes with the extra overhead of having to interpret ANSI control sequences and calculate soft line breaks, which gets extra complicated when dealing with Unicode characters.

I'm not sure if this is something that's going to be implemented natively into `bat`, but there's a couple ways you can get around it. I ended up installing a newer version of `less` with `brew install less`, but if that's not feasible for your setup, you could also just have `bat` skip the pager and let use Terminal.app's scrollback buffer instead.

**Comment by lilyball** [comment]
Is there any way I could configure bat's default pager? That way I could work around this in the Nix package by having it default to Nix-installed `less`. One option I have is to wrap bat in a script that sets `PATH` to include `less` (Nix offers this as a convenience) but then that would change the `less` that's invoked for a manual setting of `--pager="less -FR"` and I'm not sure I want to muck with that in the case where the user hasn't installed the newer `less` into their path.

I could of course patch bat's source but that doesn't seem ideal.

The simple way for bat to offer this customization is to read an env var during build time for the default pager with a fallback to `"less"`.

**Comment by sharkdp** [comment]
Thank you for the feedback.

Just to make sure I understand your idea: you would basically have the `bat` package depend on a newer version of `less` and then you want to install a version of `bat` which has a "hard-coded" reference to the new less version as a default pager?

I think that would be a great win for users and I'd love to support that. Using a build-time environment variable doesn't sound too bad to me and that should be really simple to implement.

Do we still want to fall back to "standard" `less` if the hard-coded path is not available anymore (which should not happen)?

**Comment by lilyball** [comment]
> Just to make sure I understand your idea: you would basically have the `bat` package depend on a newer version of `less` and then you want to install a version of `bat` which has a "hard-coded" reference to the new less version as a default pager?

Yes. If you provide the hook then it's trivial to express this dependency in Nix, and what's more, Nix will ensure the depended-upon version of `less` will exist on disk for as long as the `bat` that uses it exists. This works even if that `less` isn't in my `PATH`.

> Do we still want to fall back to "standard" `less` if the hard-coded path is not available anymore (which should not happen)?

I don't want bat to test if the hard-coded path is available, I want it to just use whatever I provide in lieu of hard-coding the string `less`. If I'm providing the build-time override, it's incumbent upon myself to ensure that pager actually exists.

Basically, I just want this line:

https://github.com/sharkdp/bat/blob/28266ee4418e8d4b34e0080a75463215b5387f68/src/output.rs#L54

to be replaced with something along the lines of

```diff
         let pager = pager_from_config
             .or(pager_from_env)
+            .or_else(|| option_env!("BAT_DEFAULT_PAGER").map(String::from))
             .unwrap_or_else(|| String::from("less"));
```

**Comment by lilyball** [comment]
Well I guess this part needs to be updated too

https://github.com/sharkdp/bat/blob/28266ee4418e8d4b34e0080a75463215b5387f68/src/output.rs#L63-L65

**Comment by sharkdp** [comment]
Just a quick note that I'm not personally going to work on this.

**Comment by lilyball** [comment]
I'm going to go ahead and close this because it won't solve the problem of specifying `--pager="less- RF"` in my `~/.config/bat/config`. Instead I'm just going to fix this at the Nix level by inserting Nix's `less` into `PATH` so bare references to `less` will refer to that instead.
